java/src/jmri/managers/AbstractProxyManager.java

```diff
@@ -299,29 +299,24 @@ public String createSystemName(@Nonnull String curAddress, @Nonnull String prefi
         return createSystemName(curAddress, prefix, getNamedBeanClass());
     }
 
-    public String getNextValidAddress(@Nonnull String curAddress, @Nonnull String prefix, boolean ignoreInitialExisting, char typeLetter) throws jmri.JmriException {
+    @Nonnull
+    public String getNextValidSystemName(@Nonnull NamedBean currentBean) throws JmriException {
+        int prefixLength = Manager.getSystemPrefixLength(currentBean.getSystemName());
+
+        String prefix = currentBean.getSystemName().substring(0, prefixLength);
+        char typeLetter = currentBean.getSystemName().charAt(prefixLength);
+
         for (Manager<E> m : mgrs) {
-            log.debug("NextValidAddress requested for {}", curAddress);
+            log.debug("getNextValidSystemName requested for {}", currentBean.getSystemName());
             if (prefix.equals(m.getSystemPrefix()) && typeLetter == m.typeLetter()) {
                 try {
-                    switch (typeLetter) { // use #getDefaultManager() instead?
-                        case 'T':
-                            return ((TurnoutManager) m).getNextValidAddress(curAddress, prefix, ignoreInitialExisting);
-                        case 'S':
-                            return ((SensorManager) m).getNextValidAddress(curAddress, prefix, ignoreInitialExisting);
-                        case 'L':
-                            return ((LightManager) m).getNextValidAddress(curAddress, prefix, ignoreInitialExisting);
-                        case 'R':
-                            return ((ReporterManager) m).getNextValidAddress(curAddress, prefix, ignoreInitialExisting);
-                        default:
-                            return null;
-                    }
+                    return ((NameIncrementingManager) m).getNextValidSystemName(currentBean);
```

Comment:
Is it sure that `m` is a `NameIncrementingManager`? I'm thinking of a case there some concrete managers implement it and some don't, in the same ProxyManager. For example that the LocoNet manager does but the internal manager doesn't.

Maybe an instanceof check would be good that would throw a JmriException instead of an ClassCastException. But it's only a suggestion or a question, not a request for change.

Reply:
All the (Type)Managers, e.g. all `TurnoutManager`s, descend from the same base, so they all either implement it or not; there can't be some `TurnoutManager` objects that do and some that don't.

What could go wrong is that somebody could use a `ProxyManager` for a type that doesn't inherit from `NameIncrementingManager`, which would then throw.  But that's a development-time error that they'd pick up quickly via `ClassCastException` first time through, not something worth checking every time through the code.